mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -0,0 +1,547 @@
+    "gamma" -> "inverse", "gamma" -> "identity", "gamma" -> "log"
+  )
+}
+
+/**
+ * :: Experimental ::
+ * Model produced by [[GeneralizedLinearRegression]].
+ */
+@Experimental
+@Since("2.0.0")
+class GeneralizedLinearRegressionModel private[ml] (
+    @Since("2.0.0") override val uid: String,
+    @Since("2.0.0") val coefficients: Vector,
+    @Since("2.0.0") val intercept: Double)
+  extends RegressionModel[Vector, GeneralizedLinearRegressionModel]
+  with GeneralizedLinearRegressionParams {
+
+  private lazy val familyLink = $(family) match {
+    case "gaussian" => if (isDefined(link)) Gaussian($(link)) else Gaussian("identity")
+    case "binomial" => if (isDefined(link)) Binomial($(link)) else Binomial("logit")
+    case "poisson" => if (isDefined(link)) Poisson($(link)) else Poisson("log")
+    case "gamma" => if (isDefined(link)) Gamma($(link)) else Gamma("inverse")
+  }
+
+  override protected def predict(features: Vector): Double = {
+    val eta = BLAS.dot(features, coefficients) + intercept
+    familyLink.fitted(eta)
+  }
+
+  @Since("2.0.0")
+  override def copy(extra: ParamMap): GeneralizedLinearRegressionModel = {
+    copyValues(new GeneralizedLinearRegressionModel(uid, coefficients, intercept), extra)
+      .setParent(parent)
+  }
+}
+
+/**
+ * A description of the error distribution and link function to be used in the model.
+ * @param link a link function instance
+ */
+private[ml] abstract class Family(val link: Link) extends Serializable {
+
+  /** Initialize the starting value for mu. */
+  def initialize(y: Double, weight: Double): Double
+
+  /** The variance of the endogenous variable's mean, given the value mu. */
+  def variance(mu: Double): Double
+
+  /** Weights for IRLS steps. */
+  def weights(mu: Double): Double = {
+    val x = clean(mu)
+    1.0 / (math.pow(this.link.deriv(x), 2.0) * this.variance(x))
+  }
+
+  /** The adjusted response variable. */
+  def adjusted(y: Double, mu: Double, eta: Double): Double = {
+    val x = clean(mu)
+    eta + (y - x) * link.deriv(x)
+  }
+
+  /** Linear predictor based on given mu. */
+  def predict(mu: Double): Double = this.link.link(clean(mu))
+
+  /** Fitted value based on linear predictor eta. */
+  def fitted(eta: Double): Double = clean(this.link.unlink(eta))
+
+  /** Trim the fitted value so that it will be in valid range. */
+  def clean(mu: Double): Double = mu
+
+  val epsilon: Double = 1E-16
+}
+
+/**
+ * Gaussian exponential family distribution.
+ * The default link for the Gaussian family is the identity link.
+ * @param link a link function instance
+ */
+private[ml] class Gaussian(link: Link = Identity) extends Family(link) {
+
+  override def initialize(y: Double, weight: Double): Double = {
+    if (link == Log) {
+      require(y > 0.0, "The response variable of Gaussian family with Log link " +
+        s"should be positive, but got $y")
+    }
+    y
+  }
+
+  def variance(mu: Double): Double = 1.0
+
+  override def clean(mu: Double): Double = {
```

Comment:
Here we constrict ```mu``` in valid range using a different method compared with R. In R, if ```mu``` or ```eta``` is invalid, it will  diminish ```coefficients``` until it makes ```validmu``` and ```validaeta``` passed. I think is will make convergence slowness. I'm looking forward to hear others' thought.